src/fu-dfu-firmware.c

```diff
@@ -0,0 +1,373 @@
+/*
+ * Copyright (C) 2019 Richard Hughes <richard@hughsie.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#define G_LOG_DOMAIN				"FuFirmware"
+
+#include "config.h"
+
+#include "fu-common.h"
+#include "fu-dfu-firmware.h"
+
+typedef struct {
+	guint16			 vid;
+	guint16			 pid;
+	guint16			 release;
+	guint16			 version;
+} FuDfuFirmwarePrivate;
+
+G_DEFINE_TYPE_WITH_PRIVATE (FuDfuFirmware, fu_dfu_firmware, FU_TYPE_FIRMWARE)
+#define GET_PRIVATE(o) (fu_dfu_firmware_get_instance_private (o))
+
+static void
+fu_dfu_firmware_to_string (FuFirmware *firmware, guint idt, GString *str)
+{
+	FuDfuFirmware *self = FU_DFU_FIRMWARE (firmware);
+	FuDfuFirmwarePrivate *priv = GET_PRIVATE (self);
+	fu_common_string_append_kx (str, idt, "Vid", priv->vid);
+	fu_common_string_append_kx (str, idt, "Pid", priv->pid);
+	fu_common_string_append_kx (str, idt, "Release", priv->release);
+	fu_common_string_append_kx (str, idt, "Version", priv->version);
+}
+
+/**
+ * fu_dfu_firmware_get_vid:
+ * @self: a #FuDfuFirmware
+ *
+ * Gets the vendor ID.
+ *
+ * Return value: a vendor ID, or 0xffff for unset
+ **/
+guint16
+fu_dfu_firmware_get_vid (FuDfuFirmware *self)
+{
+	FuDfuFirmwarePrivate *priv = GET_PRIVATE (self);
+	g_return_val_if_fail (FU_IS_DFU_FIRMWARE (self), 0x0);
+	return priv->vid;
+}
+
+/**
+ * fu_dfu_firmware_get_pid:
+ * @self: a #FuDfuFirmware
+ *
+ * Gets the product ID.
+ *
+ * Return value: a product ID, or 0xffff for unset
+ **/
+guint16
+fu_dfu_firmware_get_pid (FuDfuFirmware *self)
+{
+	FuDfuFirmwarePrivate *priv = GET_PRIVATE (self);
+	g_return_val_if_fail (FU_IS_DFU_FIRMWARE (self), 0x0);
+	return priv->pid;
+}
+
+/**
+ * fu_dfu_firmware_get_release:
+ * @self: a #FuDfuFirmware
+ *
+ * Gets the device ID.
+ *
+ * Return value: a device ID, or 0xffff for unset
+ **/
+guint16
+fu_dfu_firmware_get_release (FuDfuFirmware *self)
+{
+	FuDfuFirmwarePrivate *priv = GET_PRIVATE (self);
+	g_return_val_if_fail (FU_IS_DFU_FIRMWARE (self), 0x0);
+	return priv->release;
+}
+
+/**
+ * fu_dfu_firmware_get_version:
+ * @self: a #FuDfuFirmware
+ *
+ * Gets the file format version.
+ *
+ * Return value: integer, or the default 0x0100 if unset
+ **/
+guint16
+fu_dfu_firmware_get_version (FuDfuFirmware *self)
+{
+	FuDfuFirmwarePrivate *priv = GET_PRIVATE (self);
+	g_return_val_if_fail (FU_IS_DFU_FIRMWARE (self), 0x0);
+	return priv->version;
+}
+
+/**
+ * fu_dfu_firmware_set_vid:
+ * @self: a #FuDfuFirmware
+ * @vid: vendor ID, or 0xffff for unset
+ *
+ * Sets the vendor ID.
+ **/
+void
+fu_dfu_firmware_set_vid (FuDfuFirmware *self, guint16 vid)
+{
+	FuDfuFirmwarePrivate *priv = GET_PRIVATE (self);
+	g_return_if_fail (FU_IS_DFU_FIRMWARE (self));
+	priv->vid = vid;
+}
+
+/**
+ * fu_dfu_firmware_set_pid:
+ * @self: a #FuDfuFirmware
+ * @pid: product ID, or 0xffff for unset
```

Comment:
Unset?

Reply:
would `by default` be clearer?

Reply:
It's a set function, there should be no default

Reply:
Maybe I'm just being unclear. In DFU the 0xffff value means "matches any vendor" -- maybe I should just say that?

Reply:
Yes that's better